File: zig/rev.zig
const warn = @import("std").debug.warn;
const DirectAllocator = @import("std").heap.DirectAllocator;
const SegmentedList = @import("std").SegmentedList;

const side_mark = "|";
const top_bot_mark = "-----------------\n";
const empty_mark = " ";
const black_mark = "#";
const white_mark = "o";

const opening_blacks: u64 = 34628173824;
const opening_whites: u64 = 68853694464;

const left_bit: u64 = 0x8000000000000000;
const right_col: u64 = 72340172838076673; 
const left_col: u64 = 9259542123273814144;
const not_right_col: u64 = 18374403900871474942;
const not_left_col: u64 = 9187201950435737471;

pub fn printBoard(blacks: u64, whites: u64) void {
    warn(top_bot_mark);
    var i: u6 = 0;
    var c: u6 = 0;
    while (true) {
        c = i % 8;
        if (c == 0) {
            warn(side_mark);
        }
        if ((blacks << i) & left_bit != 0) {
            warn(black_mark);
        }
        else if ((whites << i) & left_bit != 0) {
            warn(white_mark);
        }
        else {
            warn(empty_mark);
        }
        if (c == 7) {
            warn("|\n");
        }
        else {
            warn(empty_mark);
        }
        if (i == 63) {
            break;
        }
        i += 1;
    }
    warn(top_bot_mark);
}

pub fn rotatePiecesCCW(pieces: u64) u64 {
    const lr: u64 = 1;
    var rotated: u64 = 0;

    var i: u6 = 0;
    while (true) {
        rotated <<= 1;

        var dx: u6 = i % 8;
        var dy: u6 = i / 8;
        var mask: u64 = lr << (56 - dx * 8 + dy);
        var set_piece: bool = (mask & pieces) != 0;

        if (set_piece) {
            rotated ^= 1;
        }

        if (i == 63) {
            break;
        }
        i += 1;
    }
    return rotated;
}

pub fn flipPiecesTB(pieces: u64) u64 {
    const row: u64 = 255;
    var flipped: u64 = 0;

    var i: u6 = 0;
    while (true) {
        flipped <<= 8;

        var mask: u64 = row << (i * 8);
        mask = mask & pieces;
        mask = mask >> (i * 8);
        flipped ^= mask;

        if (i == 7) {
            break;
        }
        i += 1;
    }
    return flipped;
}

const Player = enum(u1) {
    Black,
    White,
};

const Board = struct {
    blacks: u64,
    whites: u64,
    pub fn empties(self: Board) u64 {
        return ~(self.blacks | self.whites);
    }
    pub fn print(self: Board) void {
        printBoard(self.blacks, self.whites);
    }
    pub fn rotateCCW(self: *Board) void {
        self.blacks = rotatePiecesCCW(self.blacks);
        self.whites = rotatePiecesCCW(self.whites);
    }
    pub fn flipTB(self: *Board) void {
        self.blacks = flipPiecesTB(self.blacks);
        self.whites = flipPiecesTB(self.whites);
    }
    pub fn canonicalize(self: *Board) void {
        var blacks: [8]u64 = undefined;
        var whites: [8]u64 = undefined;
        blacks[0] = self.blacks;
        whites[0] = self.whites;
        blacks[1] = rotatePiecesCCW(blacks[0]);
        whites[1] = rotatePiecesCCW(whites[0]);
        blacks[2] = rotatePiecesCCW(blacks[1]);
        whites[2] = rotatePiecesCCW(whites[1]);
        blacks[3] = rotatePiecesCCW(blacks[2]);
        whites[3] = rotatePiecesCCW(whites[2]);
        blacks[4] = flipPiecesTB(blacks[0]);
        whites[4] = flipPiecesTB(whites[0]);
        blacks[5] = flipPiecesTB(blacks[1]);
        whites[5] = flipPiecesTB(whites[1]);
        blacks[6] = flipPiecesTB(blacks[2]);
        whites[6] = flipPiecesTB(whites[2]);
        blacks[7] = flipPiecesTB(blacks[3]);
        whites[7] = flipPiecesTB(whites[3]);
        var min_index: u32 = 0;
        var i: u32 = 1;
        while (i < 8) {
            if (blacks[i] < blacks[min_index] or
                  (blacks[i] == blacks[min_index] and whites[i] < whites[min_index])) {
                min_index = i;
            }
            i += 1;
        }
        self.blacks = blacks[min_index];
        self.whites = whites[min_index];
    }
};

const BoardList = SegmentedList(Board, 1024);
const TurnList = SegmentedList(Player, 1024);

const States = struct {
    boards: BoardList,
    turns: TurnList,
};

pub fn addChildBoards(states: *States, index: u32) u32 {
    var up_moves: u64 = undefined;
    var dn_moves: u64 = undefined;
    var lf_moves: u64 = undefined;
    var rt_moves: u64 = undefined;
    var ur_moves: u64 = undefined;
    var ul_moves: u64 = undefined;
    var dr_moves: u64 = undefined;
    var dl_moves: u64 = undefined;
    var moves: u64 = undefined;

    var mv: u64 = undefined;
    var flip: u64 = undefined;
    var tmp: u64 = undefined;
    var c1_mv: u64 = undefined;
    var c2_mv: u64 = undefined;
    var c2_run: u64 = undefined;

    var board: *Board = states.boards.at(index);
    var turn: Player = states.turns.at(index).*;

    var e: u64 = board.empties();
    var b: u64 = board.blacks;
    var w: u64 = board.whites;

    var c1: u64 = if (turn == Player.Black) b else w;
    var c2: u64 = if (turn == Player.Black) w else b;

    c1_mv = (c1 << 56);
    c2_mv = (c2 << 48);
    up_moves = c2_mv & c1_mv;
    for ([5]u6{5, 4, 3, 2, 1}) |i| {
        c1_mv = (c1 << (8 *% (i +% 1)));
        c2_mv = (c2 << (8 *% i ));
        up_moves = c2_mv & (c1_mv | up_moves);
    }
    up_moves = up_moves & e;

    c1_mv = (c1 << 56);
    c2_mv = (c2 << 48);
    dn_moves = c2_mv & c1_mv;
    for ([5]u6{5, 4, 3, 2, 1}) |i| {
        c1_mv = (c1 >> (8 *% (i +% 1)));
        c2_mv = (c2 >> (8 *% i ));
        dn_moves = c2_mv & (c1_mv | dn_moves);
    }
    dn_moves = dn_moves & e;

    warn("up moves:\n");
    printBoard(0, up_moves);
    warn("dn moves:\n");
    printBoard(0, dn_moves);

    //TODO

    var next_board = states.boards.addOne();
    var next_turn = states.turns.addOne();

    return index;
}


pub fn main() !void {
    const opening_board = Board {
        .blacks = opening_blacks,
        .whites = opening_whites,
    };

    var da = DirectAllocator.init();
    defer da.deinit();

    var states = States {
        .boards = BoardList.init(&da.allocator),
        .turns = TurnList.init(&da.allocator),
    };
    defer states.boards.deinit();
    defer states.turns.deinit();

    try states.boards.push(opening_board);
    try states.turns.push(Player.White);

    var index: u32 = 0;
    var next = addChildBoards(&states, index);

    states.boards.at(0).print();
    warn("{}\n", next);

    states.boards.at(0).blacks |= 7;
    states.boards.at(0).print();
    states.boards.at(0).rotateCCW();
    states.boards.at(0).print();
    states.boards.at(0).flipTB();
    states.boards.at(0).print();

    states.boards.at(0).canonicalize();
    states.boards.at(0).print();
}

File: nim/rev.nim
import os

const
  side_mark = "|"
  top_bot_mark = "-----------------"
  empty_mark = " "
  black_mark = "#"
  white_mark = "o"

let
  opening_blacks = 34628173824'u64
  opening_whites = 68853694464'u64

  left_bit = 0x8000000000000000'u64
  right_col = 72340172838076673'u64
  left_col = 9259542123273814144'u64
  not_right_col = 18374403900871474942'u64
  not_left_col = 9187201950435737471'u64

type
  Turn = enum
    Black
    White

  Board = object
    blacks: uint64
    whites: uint64

  States = object
    boards: seq[Board]
    turns: seq[Turn]


proc printBoard(blacks: uint64, whites: uint64) =
  echo top_bot_mark
  for i in 0..<64:
    var c = i mod 8
    if c == 0:
      stdout.write(side_mark)
    if ((blacks shl i) and left_bit) != 0:
      stdout.write(black_mark)
    elif ((whites shl i) and left_bit) != 0:
      stdout.write(white_mark)
    else:
      stdout.write(empty_mark)
    if c == 7:
      echo "|"
    else:
      stdout.write(empty_mark)
  echo top_bot_mark

proc rotatePiecesCCW(pieces: uint64): uint64 =
  let lr: uint64 = 1;
  var rotated: uint64 = 0;

  for i in 0..<64:
    rotated = rotated shl 1;

    var dx: int = i mod 8;
    var dy: int = i div 8;
    var mask: uint64 = lr shl (56 - dx * 8 + dy);
    var set_piece: bool = (mask and pieces) != 0;

    if (set_piece):
      rotated = rotated xor 1;

  return rotated

proc flipPiecesTB(pieces: uint64): uint64 =
  let row: uint64 = 255;
  var flipped: uint64 = 0;

  for i in 0..<64:
    flipped = flipped shl 8

    var mask: uint64 = row shl (i * 8)
    mask = mask and pieces
    mask = mask shr (i * 8)

    flipped = flipped xor mask

  return flipped

proc canonicalPieces(blacks: uint64, whites: uint64, canon_blacks: var uint64, canon_whites: var uint64) =
  var blacks_tfm: array[0..7, uint64]
  var whites_tfm: array[0..7, uint64]
  blacks_tfm[0] = blacks
  blacks_tfm[1] = rotatePiecesCCW(blacks_tfm[0])
  blacks_tfm[2] = rotatePiecesCCW(blacks_tfm[1])
  blacks_tfm[3] = rotatePiecesCCW(blacks_tfm[2])
  blacks_tfm[4] = flipPiecesTB(blacks_tfm[0])
  blacks_tfm[5] = flipPiecesTB(blacks_tfm[1])
  blacks_tfm[6] = flipPiecesTB(blacks_tfm[2])
  blacks_tfm[7] = flipPiecesTB(blacks_tfm[3])
  whites_tfm[0] = whites
  whites_tfm[1] = rotatePiecesCCW(whites_tfm[0])
  whites_tfm[2] = rotatePiecesCCW(whites_tfm[1])
  whites_tfm[3] = rotatePiecesCCW(whites_tfm[2])
  whites_tfm[4] = flipPiecesTB(whites_tfm[0])
  whites_tfm[5] = flipPiecesTB(whites_tfm[1])
  whites_tfm[6] = flipPiecesTB(whites_tfm[2])
  whites_tfm[7] = flipPiecesTB(whites_tfm[3])
  var min_index = 0
  for i in 0..<8:
    var blacks_lower: bool = blacks_tfm[i] < blacks_tfm[min_index]
    var blacks_equal: bool = blacks_tfm[i] == blacks_tfm[min_index]
    var whites_lower: bool = whites_tfm[i] < whites_tfm[min_index]
    if blacks_lower or (blacks_equal and whites_lower):
      min_index = i

  canon_blacks = blacks_tfm[min_index]
  canon_whites = whites_tfm[min_index]

proc print(board: Board) =
  printBoard(board.blacks, board.whites)

proc empties(board: Board): uint64 =
  return not(board.blacks or board.whites)

proc rotateCCW(board: var Board) =
  board.blacks = rotatePiecesCCW(board.blacks)
  board.whites = rotatePiecesCCW(board.whites)

proc flipTB(board: var Board) =
  board.blacks = flipPiecesTB(board.blacks)
  board.whites = flipPiecesTB(board.whites)

proc canonical(board: var Board) =
  canonicalPieces(board.blacks, board.whites, board.blacks, board.whites)




var states: States;
states.boards.add(Board(blacks: opening_blacks, whites: opening_whites))

printBoard(17, 64)
states.boards[0].print()
states.boards[0].rotateCCW()
states.boards[0].print()
states.boards[0].flipTB()
states.boards[0].print()
states.boards[0].canonical()
states.boards[0].print()


